### PR TITLE
MQTT Persistence for non-registered events

### PR DESCRIPTION
**Pull request by Praveenrajmani**
Persist events even when the client has not registered the token.

## Description
The client has to register and obtain a valid token atleast once to make the events persist.

## Motivation and Context
The events are lost if minio is started without starting the mqtt broker

## Regression
No

## How Has This Been Tested?

- Do not start the mqtt broker yet.
- Start minio with ARN and queueDir configured - we can see the ARN registered even when the mqtt broker is not yet started ( Which was responding with ARN Not found error before)
- Start sending events 

We can see the events persisting in the queueDir even if the broker is not started yet. Now, if we start the broker. We can see the events getting queued. ( We can see it in activeMQ webconsole `http://localhost:8161/admin/topics.jsp`)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by Praveenrajmani** [comment]
Since it is an urgent fix, I have vendored my own repository for now - "github.com/Praveenrajmani/paho.mqtt.golang". Have sent a PR to the upstream [https://github.com/eclipse/paho.mqtt.golang/pull/281](https://github.com/eclipse/paho.mqtt.golang/pull/281).

Will update once it gets merged.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#96ebce4e-38a5-410d-a63b-20c0fa34dd04) |
<a name="96ebce4e-38a5-410d-a63b-20c0fa34dd04"></a>
### 7250-0893c18/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32024
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 703045dd39d6:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "presignedPut(bucket_name,file_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-371fbc68639a",
    "file_name": "datafile-1-MB"
  },
  "duration": 688.54,
  "error": "Connection reset by peer",
  "status": "FAIL"
}

Executed 3 out of 13 tests successfully.
```

File: pkg/event/target/mqtt.go [hunk 1]
```diff
@@ -21,15 +21,19 @@ import (
...
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/url"
 	"path/filepath"
+	"strconv"
 	"time"
 
-	"github.com/eclipse/paho.mqtt.golang"
+	"github.com/Praveenrajmani/paho.mqtt.golang"
```
Comment (balamurugana) [inline pkg/event/target/mqtt.go]: I believe you have already sent the fix to `paho.mqtt.golang`. How about keeping this under `pkg/event/target/paho.mqtt.golang` till the PR merged in upstream?

File: pkg/event/target/mqtt.go [hunk 2]
```diff
@@ -104,7 +108,12 @@ func (target *MQTTTarget) Send(eventData event.Event) error {
 		return err
 	}
 
-	token := target.client.Publish(target.args.Topic, target.args.QoS, false, string(data))
+	var token mqtt.Token
+	if !target.client.IsConnected() {
+		token = target.client.PublishNonRegisteredEvent(target.args.Topic, target.args.QoS, false, string(data))
```
Comment (balamurugana) [inline pkg/event/target/mqtt.go]: How about keeping this logic transparent in `target.client.Publish()` than introducing new method?

File: pkg/event/target/mqtt.go [hunk 3]
```diff
@@ -140,8 +156,25 @@ func NewMQTTTarget(id string, args MQTTArgs) (*MQTTTarget, error) {
...
 	token := client.Connect()
-	if token.Wait() && token.Error() != nil {
-		return nil, token.Error()
+
+	if args.QueueDir != "" {
+		go func() {
+			// Repeat the pings until the client registers the token.
+			for {
+				if token.Wait() && token.Error() == nil {
+					fmt.Println("MQTT: Connected")
```
Comment (balamurugana) [inline pkg/event/target/mqtt.go]: remove all debug prints.

File: pkg/event/target/mqtt.go [hunk 4]
```diff
@@ -140,8 +156,25 @@ func NewMQTTTarget(id string, args MQTTArgs) (*MQTTTarget, error) {
...
+					break
+				}
+				fmt.Println("MQTT: " + token.Error().Error() + ". Retrying in " + strconv.Itoa(retryTimeout) + " seconds.")
+				time.Sleep(retryTimeout * time.Second)
+				token = client.Connect()
```
Comment (balamurugana) [inline pkg/event/target/mqtt.go]: How about connection retry with store directory transparent inside `paho.mqtt.golang`?

Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Yes the library did not have any apis exposed on retrying the connection. Their `SetReconnectInterval` did not work in our case. Shall we add an API (say, `RetryConnect`) and move this inside that ?

Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: I thought this is more client specific and did not have a plan to move it inside the library

File: pkg/event/target/mqtt.go (hunk 3 above)
Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: This was intentional to let the user know that we are retrying. Should i remove this ?

File: pkg/event/target/mqtt.go (hunk 1 above)
Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Will do :+1:

Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Done

File: pkg/event/target/mqtt.go (hunk 2 above)
Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Done :+1:

File: pkg/event/target/mqtt.go (hunk 3 above)
Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Done

File: pkg/event/target/mqtt.go (hunk 4 above)
Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Done :+1:

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
Please update https://github.com/eclipse/paho.mqtt.golang/pull/281

File: pkg/event/target/mqtt.go [hunk 5]
```diff
@@ -135,13 +144,17 @@ func NewMQTTTarget(id string, args MQTTArgs) (*MQTTTarget, error) {
 		AddBroker(args.Broker.String())
 
 	if args.QueueDir != "" {
-		options = options.SetStore(mqtt.NewFileStore(args.QueueDir))
+		options = options.SetStore(mqtt.NewFileStore(args.QueueDir)).SetIgnoreAndRetryConnection(true).SetMaxReconnectInterval(retryTimeout * time.Second)
```
Comment (balamurugana) [inline pkg/event/target/mqtt.go]: Move `5 * time.Second` as default inside `paho.mqtt.golang` and this option will be avoided here.

Comment (balamurugana) [reply pkg/event/target/mqtt.go]: `SetConnectionRetry()` is better than `SetIgnoreAndRetryConnection()`

Comment (Praveenrajmani) [reply pkg/event/target/mqtt.go]: Done :+1:

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#ae80f4cc-1b47-4f10-a6cf-5581a1f39e1d) |
| mint-compression-dist-xl.sh | :x: [more...](#cf170603-15db-4176-bac7-67e675624224) |
| mint-fs.sh | :x: [more...](#bb13f780-4106-4ed5-b7b3-a4d23dac61ff) |
| mint-dist-xl.sh | :x: [more...](#eedbfcfe-5b95-4160-8d8c-27b386101f86) |
<a name="bb13f780-4106-4ed5-b7b3-a4d23dac61ff"></a>
### 7250-184c120/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30026
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4b510077ce2f:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-6ytexp96idkp4z",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 323,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.61:30026/aws-sdk-go-test-6ytexp96idkp4z: dial tcp 72.28.97.61:30026: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="eedbfcfe-5b95-4160-8d8c-27b386101f86"></a>
### 7250-184c120/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31164
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 97e873809fe6:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ppne5cimvp9nle",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 349,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:31164/aws-sdk-go-test-ppne5cimvp9nle: dial tcp 72.28.97.57:31164: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="ae80f4cc-1b47-4f10-a6cf-5581a1f39e1d"></a>
### 7250-184c120/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32107
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9d332aa99e44:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-kd1tw6bibx6d5t",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 366,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.54:32107/aws-sdk-go-test-kd1tw6bibx6d5t: dial tcp 72.28.97.54:32107: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="cf170603-15db-4176-bac7-67e675624224"></a>
### 7250-184c120/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31892
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 16202d8401ca:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-4wwt1jrjhcv1nv",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 384,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:31892/aws-sdk-go-test-4wwt1jrjhcv1nv: dial tcp 72.28.97.57:31892: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```

**Comment by balamurugana** [comment]
`pkg/event/target/paho.mqtt.golang`  needs to be ignored in `gofmt` check.

**Comment by harshavardhana** [comment]
We have decided to do a generic approach to persistence for all targets with a configurable parameter @Praveenrajmani - will let you know the config changes needed...

**Comment by harshavardhana** [comment]
> We have decided to do a generic approach to persistence for all targets with a configurable parameter @Praveenrajmani - will let you know the config changes needed...

The generic approach is to add two new fields to each target

```
"queueDir": ""
"queueLimit": N,
```

QueueDir similar to MQTT feature stores the payload - each target should implement this perhaps a common Store interface which each target should comply and use in Send(). But we will not use the MQTT store feature here instead use this new feature at our layer.

QueueDir if set will persist to the directory, if not persist in memory.  QueueLimit is to limit to the number of messages persisted at queueDir or memory. QueueLimit if not set we should limit to 10,000 messages by default. 

QueueDir should always save the event message unless and until it's delivered successfully if the success is received upon delivery the message is deleted on our end. In any situation, the message couldn't be delivered we will retry from the persistent directory. 

NOTE: we are not planning to provide ordered delivery guarantees of messages, ordering is already possible on the other end by looking at timestamps per message.

Each QueueDir should be unique for a target and cannot share the same queueDir - all servers use this directory to store locally, we are not going to erasure code these messages while persisting so there is no reason to use object layer.

Additionally no need to update the config version, new fields will appear automatically upon upgrade and MQTT users shouldn't see any difference.

**Comment by harshavardhana** [comment]
Now support for each target can come in separately from MQTT point of view, we can add such support right now, rest of the targets can simply use a memory store for now since they don't need to have queueDir support right away and MQTT feature is not blocking other implementations.

**Comment by Praveenrajmani** [comment]
As we decided to fix this in a generic approach. This change in the PR is not required. Will close this and open a new PR.

**Pull request closed without merging**